### PR TITLE
validator reward by encoding proposer frequency into epoch header

### PR DESCRIPTION
**Pull request by ping-ke**
Functions:
1. increase the proposed block count for the state when a block has been applied in the tendermint consensus.
2. add state rlp to epoch header extraData and reset state
3. relayer: adding header number to submit function params 

Test:
1. integration test
2. use the proposed state checker to verify epoch header extraData

### CODE REVIEW

**Review by qizhou — COMMENTED** [review]
In general, it looks good.  Just a question regarding performance.

File: consensus/tendermint/adapter/store.go [hunk 1]
```diff
@@ -321,3 +343,31 @@ func (s *Store) MakeBlock(
 
 	return &types.FullBlock{Block: block, LastCommit: commit}
 }
+
+func (s *Store) getEpochState(lastEpochHeight uint64, startHeader *types.Header) ([]byte, error) {
+	m := make(map[common.Address]uint32)
+	header := startHeader
+	for header.Number.Uint64() > lastEpochHeight {
```
[the remark this answers is not in the document]
Comment (ping-ke) [reply consensus/tendermint/adapter/store.go]: it takes 18s to collect 100800 blocks coinbase.

**Comment by qizhou** [comment]
What is the test plan?

**Comment by ping-ke** [comment]
1. integration test 
2. use the proposed state checker to verify epoch header extraData

File: consensus/tendermint/adapter/store.go [hunk 2]
```diff
@@ -39,7 +42,25 @@ func NewStore(
...
+		makeBlock:        makeBlock,
+		gov:              gov,
+		mux:              mux,
+		state:            make(map[common.Address]uint32),
+	}
+
+	t := time.Now()
+	header := chain.CurrentHeader()
+	lastEpochHeight := header.Number.Uint64() - header.Number.Uint64()%config.Epoch
+	for header != nil && header.Number.Uint64() >= lastEpochHeight {
```
Comment (qizhou) [inline consensus/tendermint/adapter/store.go:58]: Should the condition be `header.number.Unit64() > lastEpochHeight`?  `lastEpochHeight` should be block that contains the validator change, but the proposer of the block should not be counted towards the last epoch?

File: consensus/tendermint/adapter/store.go [hunk 3]
```diff
@@ -102,22 +123,24 @@ func (s *Store) ValidateBlock(state pbft.ChainState, block *types.FullBlock, com
 	}
 
 	validators, powers := []common.Address{}, []uint64{}
-	if header.Number.Uint64()%s.config.Epoch == 0 {
-		epochId := header.Number.Uint64() / s.config.Epoch
+	height := header.Number.Uint64()
+	if height%s.config.Epoch == 0 {
+		epochId := height / s.config.Epoch
+		extraData := header.Extra
```
Comment (qizhou) [inline consensus/tendermint/adapter/store.go:129]: I feel a comment about extraData format should be given here.

**Review by qizhou — COMMENTED** [review]
I feel `proposedstate` is not a good name.  Perhaps `proposers_checker` may be better (or if you have a better name).

File: consensus/tendermint/adapter/store.go [hunk 4]
```diff
@@ -288,7 +330,17 @@ func (s *Store) MakeBlock(
 	header.TimeMs = timestampMs
 	header.LastCommitHash = commit.Hash()
 
-	if height%s.config.Epoch == 0 {
+	if height%s.config.Epoch == 0 && height > 0 {
```
Comment (qizhou) [inline consensus/tendermint/adapter/store.go:333]: In which case that `height == 0`?

**Review by qizhou — APPROVED** [review]
In general, it looks good.  One comment is whether we should add a version field in extra data in case we want to have a completely different format.

**Pull request merged**